doc/copp/Copp_Manager_Redesign_test_plan.md

```diff
@@ -0,0 +1,134 @@
+
+### Scope
+
+The test is to verify the Copp Manager redesign logic for installing or uninstalling trap, when there are different configuration combinations for always_enabled and the feature of the corresponding trap.   
+
+### Scale / Performance
+
+No scale/performance test involved in this test plan
+
+### Related **DUT** CLI commands
+
+No CLI
+
+### Related DUT configuration files
+
+The traps configuration are saved to  /etc/sonic/copp_cfg.json.
+
+```
+    "COPP_TRAP": {
+    "bgp": {
+        "trap_ids": "bgp,bgpv6",
+        "trap_group": "queue4_group1"
+    },
+    "arp": {
+        "trap_ids": "arp_req,arp_resp,neigh_discovery",
+        "trap_group": "queue4_group2",
+        "always_enabled": "true"
+    },
+    "udld": {
+	    "trap_ids": "udld",
+	    "trap_group": "queue4_group3",
+         "always_enabled": "true"
+    },
+    "ip2me": {
+	    "trap_ids": "ip2me",
+	    "trap_group": "queue1_group1",
+         "always_enabled": "true"
+    },
+    "lacp": {
+	    "trap_ids": "lacp",
+	    "trap_group": "queue4_group1",
+         "always_enabled": "true"
+    }
+}
+```
+### Supported topology
+The test will be supported on ptf32, ptf64, t1 and t2.
+
+
+## Test cases
+### Test cases #1 -  Verify 4 traps'default configuration
+1. Verify the 4 traps(arp, lacp, udld, ip2me) are installed  by sending traffic with the default config
+
+### Test cases #2 - Add a new trap
+#### Pre-condition: The tested trap should be uninstalled on dut.
+1. Set always_enabled of bgp to true with following cmds:
+   
+  + generate copp config file
+```  
+   cat copp_bgp.json
+      {
+       "COPP_TRAP": {
+           "bgp": {
+           "always_enabled": "true"
+            }
+        }
+   }
+```
+   + config load copp_bgp.json -y
+   
+2. Verify the corresponding trap status is installed by sending traffic    
+3. Recover the config
+
+   
+### Test cases #3 - Remove a trap by removing entry from feature table
+#### Pre-condition: The tested trap has been installed on dut,  with always_enable true and the corresponding entry enabled.
+Take bgp trap as example:
+1. Set always_enabled of bgp to false (refer to test #2)
+2. Remove bgp entry from feature table with CMD
+   redis-cli -n 4 hdel "FEATURE|bgp" "auto_restart" "has_global_scope" "has_per_asic_scope" "has_timer" "high_mem_alert" "state" 
+3. Verify the tested trap is uninstalled  by sending traffic
+5. Recover the config   
+
+
+### Test cases #4 - Remove a trap by disabling feature table
+Take bgp trap as an example:
+1. Set always_enabled of bgp to false (refer to test #2)
+2. Enable bgp in feature table: config feature state bgp enabled    
+3. Verify the tested trap is installed by sending traffic
+4. Disable bgp in feature table: config feature state bgp disabled 
+5. Verify the tested trap is uninstalled by sending traffic
+6. Recover the config
+
+
+   
+### Test cases #5 - Verify trap configuration is saved or not after reboot(reboot, fast-reboot, warm-reboot)
```

Comment:
I think, one more case to verify is the sflow example where the entry is present in copp.json, but not in feature table. It should not be installed until the feature is enabled

Reply:
@prsunny I believe Test case 4 should cover the scenario and our copp swss UT handles it. The reason we cannot test sflow in sonic mgmt is if we disable the feature, sampling itself will be disabled and so we cannot assume that packets are not lifted because of trap removal or sampling itself is disabled. That's why we use BGP to test this usecase.

Reply:
I meant sflow as an example. Just want to make sure the sequence is covered. But as you mentioned, Test 4 seems to cover